open-metadata-implementation/adapters/open-connectors/governance-daemon-connectors/open-lineage-connectors/janus-connector/src/main/java/org/odpi/openmetadata/openconnectors/governancedaemonconnectors/openlineageconnectors/janusconnector/GraphStore.java

```diff
@@ -0,0 +1,8 @@
+/* SPDX-License-Identifier: Apache-2.0 */
+/* Copyright Contributors to the ODPi Egeria project. */
+package org.odpi.openmetadata.openconnectors.governancedaemonconnectors.openlineageconnectors.janusconnector;
+
+public interface GraphStore {
```

Comment:
This interface should be located in the open lineage services api module.  It is an interface that other graph stores need to implement if JanusGraph is to be replaced.

It would help if it had a more meaningful name - such as HistoricalOpenLineageGraphStore.

In addition, to allow you to make changes to the API over time, you should have a base connector class and a base connector provider class in the same package.  Then the JanusGraph implementation would extend the base class.

Reply:
Ok I am trying to understand this so I want to ask. Does the BaseConnector class in Open lineage will have any implementation or just overriding the methods , I assume it will not so the methods will be empty right?. So then why can't I build a generic connector under connectors that will fetch the connector that is needed based on the configuration.

Reply:
The connector interface and base class belong to the "framework module" - that is the module that the connector implementation needs to plug in to.  In this case the open lineage services is the "framework module" and so it specifies the API that it needs connectors to implement in order to support its lineage store.  The base class allows changes to be made to this API without needing to change any existing connector implementations (the base class provides the default implementation of any new methods).

The base class typcally begins reasonably empty - it may have useful utilities defined that are useful to all connector implementations.

If the API and the base class are in the open lineage services API module then the open lineage services have no dependencies on any connector implementation.  The connector implementations can be swapped in and out without any impact on the open lineage services module.

Each connector implement is then dependent on the open lineage services API module.  It extends the connector base class.

The open lineage services does not load the connector implementation.  It uses the **ConnectorBroker** to load the connector implementation requested in the Connection object for the connector.  This Connection object is stored in the configuration document for the open lineage services.

If the API and base class of the connector is in the same module as the janus graph connector then the open lineage services are dependent on a connector implementation.  This is the wrong way around and defeats the purpose of using connectors.

We need to use connectors because different vendors may need to change the implementation of our stores, event busses etc so that Eggeria will run on their platform.